clients/src/main/java/org/apache/kafka/clients/producer/internals/BufferPool.java

```diff
@@ -109,78 +112,76 @@ public ByteBuffer allocate(int size, long maxTimeToBlockMs) throws InterruptedEx
                 // we have enough unallocated or pooled memory to immediately
                 // satisfy the request
                 freeUp(size);
+                ByteBuffer allocatedBuffer = allocateByteBuffer(size);
```

Comment:
Good point. I think it is still worth keeping the optimization, although typically the the producer will only allocate poolable batch size, so the actual memory allocation should not happen very often.